### PR TITLE
Add In Memory Db

### PR DESCRIPTION
**Pull request by zander-xyz**
The idea is that we should be able to use dependency injection to swap out the backend of the db interface. It's currently designed to be compatible with an in memory and rocksdb backend. Will tackle interface with our custom trie db when it comes

### CODE REVIEW

File: include/monad/trie/rocks_writer.hpp [hunk 1]
```diff
@@ -7,51 +7,50 @@
...
-    }
-
-    void write()
-    {
-        auto const res = db->Write(rocksdb::WriteOptions{}, &batch);
-        MONAD_ROCKS_ASSERT(res);
-        batch.Clear();
-    }
-};
+// struct RocksWriter
```
Comment (rgarc) [inline include/monad/trie/rocks_writer.hpp:16]: Is there a reason to keep this commented code lying around as opposed to just deleting it?

File: include/monad/db/trie_db.hpp [hunk 2]
```diff
@@ -0,0 +1,369 @@
...
+        }
+
+        lc.prev();
+
+        using namespace std::placeholders;
+
+        auto const left =
+            lc.key()
+                .transform(&TCursor::Key::path)
+                .transform(std::bind(trie::longest_common_prefix_size, _1, lk));
```
Comment (rgarc) [inline include/monad/db/trie_db.hpp]: small nit: is there a reason to use `std::bind` and `std::placeholders` anymore instead of lambdas? so something like this
```diff
diff --git a/include/monad/db/trie_db.hpp b/include/monad/db/trie_db.hpp
index ca8a627..3496ee2 100644
--- a/include/monad/db/trie_db.hpp
+++ b/include/monad/db/trie_db.hpp
@@ -93,12 +93,10 @@ struct TrieDB
 
         lc.prev();
 
-        using namespace std::placeholders;
-
         auto const left =
             lc.key()
                 .transform(&TCursor::Key::path)
-                .transform(std::bind(trie::longest_common_prefix_size, _1, lk));
+                .transform([&lk](const trie::Nibbles& nibbles) { return trie::longest_common_prefix_size(nibbles, lk); });
 
         lc.next();
         MONAD_DEBUG_ASSERT(lc.key().transform(&TCursor::Key::path) == lk);
@@ -107,7 +105,7 @@ struct TrieDB
         auto const right =
             lc.key()
                 .transform(&TCursor::Key::path)
-                .transform(std::bind(trie::longest_common_prefix_size, lk, _1));
+                .transform([&lk](const trie::Nibbles& nibbles) { return trie::longest_common_prefix_size(lk, nibbles); });
 
         auto const tk =
             !left.has_value() && !right.has_value()
```

File: src/monad/trie/test/helpers.hpp [hunk 3]
```diff
@@ -15,213 +15,212 @@
...
-    {
-        trie_.process_updates(updates);
-
-        writer_.write();
-        leaves_cursor_.take_snapshot();
-        trie_cursor_.take_snapshot();
-    }
-};
-
+// TODO: these no longer compile due to interface changes. Fix when we
```
Comment (rgarc) [inline src/monad/trie/test/helpers.hpp:18]: ditto as the other commented chunk of code

Comment (zander-xyz) [reply src/monad/trie/test/helpers.hpp:18]: We likely need this code again (albeit, fixed up), but I think the actual fix-up belongs in a separate PR, when we add in rocksdb support.

File: include/monad/db/state.hpp [hunk 4]
```diff
@@ -207,8 +207,8 @@ struct State
 
     void commit()
     {
-        accounts_.commit_all_merged();
         storage_.commit_all_merged();
+        accounts_.commit_all_merged();
```
Comment (sheapwilliams) [inline include/monad/db/state.hpp:214]: Why this change?

File: include/monad/db/trie_db.hpp [hunk 5]
```diff
@@ -0,0 +1,369 @@
...
+#include <map>
+#include <monad/core/account.hpp>
+#include <monad/core/assert.h>
+#include <monad/db/config.hpp>
+#include <monad/rlp/decode_helpers.hpp>
+#include <monad/trie/in_memory_comparator.hpp>
+#include <monad/trie/in_memory_cursor.hpp>
+#include <monad/trie/in_memory_writer.hpp>
+#include <monad/trie/trie.hpp>
+#include <variant>
```
Comment (sheapwilliams) [inline include/monad/db/trie_db.hpp]: include ordering? `variant` and `map` unused?  include `<unordered_map>`?

File: include/monad/db/trie_db.hpp [hunk 6]
```diff
@@ -0,0 +1,369 @@
...
+        trie::Trie<cursor_t, writer_t> trie;
+
+        std::unordered_map<address_t, std::optional<Account>> updates;
+
+        Accounts()
+            : leaves_cursor{leaves_storage}
+            , leaves_writer{leaves_storage}
+            , trie_cursor{trie_storage}
+            , trie_writer{trie_storage}
+            , trie{leaves_cursor, trie_cursor, leaves_writer, trie_writer}
```
Comment (sheapwilliams) [inline include/monad/db/trie_db.hpp]: Are some of these members left uninitialized? `leaves_storage` and `trie_storage`, in particular? And `updates`?  Is that all ok?

File: include/monad/db/trie_db.hpp [hunk 7]
```diff
@@ -0,0 +1,369 @@
...
+
+        Storage()
+            : leaves_cursor{leaves_storage}
+            , leaves_writer{leaves_storage}
+            , trie_cursor{trie_storage}
+            , trie_writer{trie_storage}
+            , trie{leaves_cursor, trie_cursor, leaves_writer, trie_writer}
+        {
+        }
+    } storage;
```
Comment (sheapwilliams) [inline include/monad/db/trie_db.hpp]: How do the `accounts` member and the `storage` member differ?  Is is just the internal `updates` member?  They look so similar...Is there a generalization of these things?  Something that helps me understand the difference?

File: include/monad/db/state.hpp (hunk 4 above)
Comment (zander-xyz) [reply include/monad/db/state.hpp:214]: Discussed over zoom, will place a comment here for clarity.

File: include/monad/db/trie_db.hpp (hunk 5 above)
Comment (zander-xyz) [reply include/monad/db/trie_db.hpp]: Yep, I'll reorder these according to the below.

monad
external
stdlib

File: include/monad/db/trie_db.hpp (hunk 6 above)
Comment (zander-xyz) [reply include/monad/db/trie_db.hpp]: I can be more explicit here for clarity.

File: include/monad/db/trie_db.hpp (hunk 7 above)
Comment (zander-xyz) [reply include/monad/db/trie_db.hpp]: Discussed over zoom. Yes, a large part of these classes is just plumbing, with the main difference being the updates.
